internal/upload/oci/upload.go

```diff
@@ -0,0 +1,186 @@
+		if err != nil {
+			return "", err
+		}
+	}
+	// clean up the object even if we fail
+	defer c.deleteObjectFromBucket(objectName, bucketName, namespace)
+
+	imageID, err := c.create(objectName, bucketName, namespace, compartmentID)
+	if err != nil {
+		return "", fmt.Errorf("failed to create a custom image using object '%s' bucket '%s' in namespace '%s': %w",
+			objectName,
+			bucketName,
+			namespace,
+			err)
+	}
+	err = c.GrantImageAccess(imageID, userID, compartmentID)
+	if err != nil {
+		return "", fmt.Errorf("failed to grant image ID '%s' access to user %s in compartment %s: ",
+			imageID,
+			userID,
+			compartmentID)
+	}
+
+	return imageID, nil
+}
+
+// Create creates an image from the storageObjectName stored in the bucketName.
+// The result is an image ID or an error if the operation failed.
+func (c Client) create(storageObjectName, bucketName, namespace, compartmentID string) (string, error) {
+	request := core.CreateImageRequest{
+		CreateImageDetails: core.CreateImageDetails{
+			CompartmentId: common.String(compartmentID),
+			ImageSourceDetails: core.ImageSourceViaObjectStorageTupleDetails{
+				BucketName:             common.String(bucketName),
+				NamespaceName:          common.String(namespace),
+				ObjectName:             common.String(storageObjectName),
+				OperatingSystem:        common.String("RHEL/Fedora"), //TODO this is just to experiment with the UI - can be removed
+				OperatingSystemVersion: nil,
+				SourceImageType:        "",
+			},
+		},
+	}
+
+	image, err := c.computeClient.CreateImage(context.Background(), request)
+	if err != nil {
+		return "", fmt.Errorf("failed to create an image from storage object: %w", err)
+	}
+	return *image.Id, nil
+}
+
+type ClientParams struct {
+	User        string
+	Region      string
+	Tenancy     string
+	PrivateKey  string
+	Fingerprint string
+}
+
+type ociClient struct {
+	storageClient  objectstorage.ObjectStorageClient
+	identityClient identity.IdentityClient
+	computeClient  core.ComputeClient
+}
+
+func (c Client) GetBucket(bucketName, namespace string) (objectstorage.Bucket, error) {
+	res, err := c.storageClient.GetBucket(context.Background(), objectstorage.GetBucketRequest{
+		NamespaceName:   common.String(namespace),
+		BucketName:      common.String(bucketName),
+		RequestMetadata: common.RequestMetadata{},
+	})
+	if err != nil {
+		return objectstorage.Bucket{}, err
+	}
+	return res.Bucket, nil
+}
+
+// GrantImageAccess grants the user the needed permissions to launch an instance
+// from the imageID in the designated compartment
+func (c Client) GrantImageAccess(imageID string, user string, compartmentID string) error {
+	// TBD
+	// should we support granting access to a group? oci have an api to grant
+	// access to groups or any-user (all users in tenancy) rather than specific user.
+	// also, not all regions support identity domains, i.e they'll need differnt method to
+	// allow access.
```

Comment:
How much work would it be to support both?  Would it need separate options, or could we have a single ID of some sort that would work for either a user or a group based on its type?

Reply:
single ID for a user is not supported. Mayeb if ID is supplied then use that as the group ID, otherwise any-user?